homeassistant/components/dsmr/sensor.py

```diff
@@ -137,44 +143,70 @@ def update_entities_telegram(telegram):
-        while hass.state != CoreState.stopping:
+        while True:
             # Start DSMR asyncio.Protocol reader
             try:
                 transport, protocol = await hass.loop.create_task(reader_factory())
-            except (
-                serial.serialutil.SerialException,
-                ConnectionRefusedError,
-                TimeoutError,
-            ):
-                # Log any error while establishing connection and drop to retry
-                # connection wait
-                _LOGGER.exception("Error connecting to DSMR")
-                transport = None
 
-            if transport:
-                # Register listener to close transport on HA shutdown
-                stop_listener = hass.bus.async_listen_once(
-                    EVENT_HOMEASSISTANT_STOP, transport.close
-                )
+                if transport:
+                    # Register listener to close transport on HA shutdown
+                    stop_listener = hass.bus.async_listen_once(
+                        EVENT_HOMEASSISTANT_STOP, transport.close
+                    )
+
+                    # Wait for reader to close
+                    await protocol.wait_closed()
 
-                # Wait for reader to close
-                await protocol.wait_closed()
+                if hass.state == CoreState.stopping:
+                    return
 
-            if hass.state != CoreState.stopping:
                 # Unexpected disconnect
                 if transport:
                     # remove listener
                     stop_listener()
 
+                transport = None
+                protocol = None
+
                 # Reflect disconnect state in devices state by setting an
                 # empty telegram resulting in `unknown` states
                 update_entities_telegram({})
 
                 # throttle reconnect attempts
                 await asyncio.sleep(config[CONF_RECONNECT_INTERVAL])
 
+            except (serial.serialutil.SerialException, OSError):
+                # Log any error while establishing connection and drop to retry
+                # connection wait
+                _LOGGER.exception("Error connecting to DSMR")
+                transport = None
+                protocol = None
+            except CancelledError:
+                if stop_listener:
+                    stop_listener()
+
+                if transport:
+                    transport.close()
+
+                if protocol:
+                    await protocol.wait_closed()
+
+                return
+
     # Can't be hass.async_add_job because job runs forever
-    hass.loop.create_task(connect_and_reconnect())
+    task = hass.loop.create_task(connect_and_reconnect())
+
+    # Save the task to be able to cancel it when unloading
+    hass.data[DOMAIN][entry.entry_id][DATA_TASK] = task
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
```

Comment:
can we keep the platform setup above entity? just makes it clearer its handling such vs just entries. also creates less movement in the diffs usually.